Failing to build

I was going to try this module out, but it fails to build on my macOS Catalina machine with the following error when trying a `yarn add`:

```
error /path/to/project/node_modules/posix-semaphore: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: /path/to/project/node_modules/posix-semaphore
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@7.1.2
gyp info using node@15.3.0 | darwin | x64
gyp info find Python using Python version 2.7.16 found at "/System/Library/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python"
gyp info spawn /System/Library/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python
gyp info spawn args [
gyp info spawn args   '/usr/local/Cellar/node/15.3.0/libexec/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/path/to/project/node_modules/posix-semaphore/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/local/Cellar/node/15.3.0/libexec/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/erin/Library/Caches/node-gyp/15.3.0',
gyp info spawn args   '-Dnode_gyp_dir=/usr/local/Cellar/node/15.3.0/libexec/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/Users/erin/Library/Caches/node-gyp/15.3.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/path/to/project/node_modules/posix-semaphore',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.'
gyp info spawn args ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  CXX(target) Release/obj.target/addon/srcs/semaphore.o
../srcs/semaphore.cc:25:9: warning: 'sem_getvalue' is deprecated [-Wdeprecated-declarations]
    if (sem_getvalue(this->semaphore, &sval)) sval = -1;
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/semaphore.h:54:56: note: 'sem_getvalue' has been explicitly marked deprecated here
int sem_getvalue(sem_t * __restrict, int * __restrict) __deprecated;
                                                       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:187:40: note: expanded from macro '__deprecated'
#define __deprecated    __attribute__((__deprecated__))
                                       ^
../srcs/semaphore.cc:54:38: error: too few arguments to function call, single argument 'context' was not specified
  constructor.Reset(tpl->GetFunction());
                    ~~~~~~~~~~~~~~~~ ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8.h:6497:3: note: 'GetFunction' declared here
  V8_WARN_UNUSED_RESULT MaybeLocal<Function> GetFunction(
  ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8config.h:431:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../srcs/semaphore.cc:55:73: error: too few arguments to function call, single argument 'context' was not specified
  exports->Set(Nan::New("Semaphore").ToLocalChecked(), tpl->GetFunction());
                                                       ~~~~~~~~~~~~~~~~ ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8.h:6497:3: note: 'GetFunction' declared here
  V8_WARN_UNUSED_RESULT MaybeLocal<Function> GetFunction(
  ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8config.h:431:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../srcs/semaphore.cc:83:34: error: too few arguments to function call, single argument 'isolate' was not specified
  strict = info[1]->BooleanValue();
           ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8.h:2904:3: note: 'BooleanValue' declared here
  bool BooleanValue(Isolate* isolate) const;
  ^
../srcs/semaphore.cc:84:33: error: too few arguments to function call, single argument 'isolate' was not specified
  debug = info[2]->BooleanValue();
          ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8.h:2904:3: note: 'BooleanValue' declared here
  bool BooleanValue(Isolate* isolate) const;
  ^
../srcs/semaphore.cc:85:34: error: too few arguments to function call, single argument 'isolate' was not specified
  silent = info[3]->BooleanValue();
           ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8.h:2904:3: note: 'BooleanValue' declared here
  bool BooleanValue(Isolate* isolate) const;
  ^
../srcs/semaphore.cc:86:42: error: too few arguments to function call, single argument 'isolate' was not specified
  retry_on_eintr = info[4]->BooleanValue();
                   ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8.h:2904:3: note: 'BooleanValue' declared here
  bool BooleanValue(Isolate* isolate) const;
  ^
../srcs/semaphore.cc:87:58: error: too few arguments to function call, single argument 'context' was not specified
  value = !info[5]->IsUndefined()? info[5]->IntegerValue(): 1;
                                   ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8.h:2909:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8config.h:431:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../srcs/semaphore.cc:88:47: error: too few arguments to function call, single argument 'context' was not specified
  v8::String::Utf8Value str(info[0]->ToString());
                            ~~~~~~~~~~~~~~~~~ ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8.h:2856:3: note: 'ToString' declared here
  V8_WARN_UNUSED_RESULT MaybeLocal<String> ToString(
  ^
/Users/erin/Library/Caches/node-gyp/15.3.0/include/node/v8config.h:431:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../srcs/semaphore.cc:114:9: warning: 'sem_getvalue' is deprecated [-Wdeprecated-declarations]
    if (sem_getvalue(obj->semaphore, &sval)) sval = -1;
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/semaphore.h:54:56: note: 'sem_getvalue' has been explicitly marked deprecated here
int sem_getvalue(sem_t * __restrict, int * __restrict) __deprecated;
                                                       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:187:40: note: expanded from macro '__deprecated'
#define __deprecated    __attribute__((__deprecated__))
                                       ^
../srcs/semaphore.cc:144:9: warning: 'sem_getvalue' is deprecated [-Wdeprecated-declarations]
    if (sem_getvalue(obj->semaphore, &sval)) sval = -1;
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/semaphore.h:54:56: note: 'sem_getvalue' has been explicitly marked deprecated here
int sem_getvalue(sem_t * __restrict, int * __restrict) __deprecated;
                                                       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:187:40: note: expanded from macro '__deprecated'
#define __deprecated    __attribute__((__deprecated__))
                                       ^
../srcs/semaphore.cc:165:9: warning: 'sem_getvalue' is deprecated [-Wdeprecated-declarations]
    if (sem_getvalue(obj->semaphore, &sval)) sval = -1;
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/semaphore.h:54:56: note: 'sem_getvalue' has been explicitly marked deprecated here
int sem_getvalue(sem_t * __restrict, int * __restrict) __deprecated;
                                                       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:187:40: note: expanded from macro '__deprecated'
#define __deprecated    __attribute__((__deprecated__))
                                       ^
../srcs/semaphore.cc:180:9: warning: 'sem_getvalue' is deprecated [-Wdeprecated-declarations]
    if (sem_getvalue(obj->semaphore, &sval)) sval = -1;
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/semaphore.h:54:56: note: 'sem_getvalue' has been explicitly marked deprecated here
int sem_getvalue(sem_t * __restrict, int * __restrict) __deprecated;
                                                       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:187:40: note: expanded from macro '__deprecated'
#define __deprecated    __attribute__((__deprecated__))
                                       ^
5 warnings and 8 errors generated.
make: *** [Release/obj.target/addon/srcs/semaphore.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/Cellar/node/15.3.0/libexec/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (node:events:376:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:277:12)
gyp ERR! System Darwin 19.6.0
gyp ERR! command "/usr/local/Cellar/node/15.3.0/bin/node" "/usr/local/Cellar/node/15.3.0/libexec/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /path/to/project/node_modules/posix-semaphore
gyp ERR! node -v v15.3.0
gyp ERR! node-gyp -v v7.1.2
```

Thought you might appreciate a heads up. Let me know if you need any further info about my system.

**EDIT**: Updated error output after fixing macOS command line tools warnings.

--- a/srcs/semaphore.cc
+++ b/srcs/semaphore.cc
@@ -86,8 +86,12 @@ void Semaphore::New(const Nan::FunctionCallbackInfo<v8::Value>& info)
 
   size_t str_len;
   str_len = str.length();
-  if (str_len >= SEMSIZE - 1 || str_len <= 0)
-    return Nan::ThrowError("Semaphore() : first argument's length must be < 255 && > 0");
+  if (str_len >= SEMSIZE || str_len <= 0)
+  {
+    char msg[100];
+    sprintf(msg, "Semaphore(): first argument's length must be < %i && > 0", SEMSIZE);
+    return Nan::ThrowError(msg);
+  }
 
   Semaphore* obj = new Semaphore(buf, str_len, strict, debug, silent, retry_on_eintr, value);
   obj->Wrap(info.This());

--- a/srcs/semaphore.cc
+++ b/srcs/semaphore.cc
@@ -1,29 +1,17 @@
 #include "semaphore.h"
+#include "typeinfo"
 
 Nan::Persistent<v8::Function> Semaphore::constructor;
 
 Semaphore::Semaphore(char buf[], size_t buf_len, bool strict, bool debug, bool silent, bool retry_on_eintr, unsigned int value /*= 1*/)
 {
-  size_t  i;
-
-  i = 0;
-  while (i < buf_len)
-  {
-    this->sem_name[i] = buf[i];
-    i++;
-  }
-  this->sem_name[buf_len] = 0;
+  strcpy(this->sem_name, buf);
   this->semaphore = sem_open(this->sem_name, O_CREAT, 0644, value);
-  if(this->semaphore == SEM_FAILED)
+  if (this->semaphore == SEM_FAILED)
   {
     this->closed = 1;
-    Nan::ThrowError("could not create semaphore : sem_open failed");
+    Nan::ThrowError("Could not create semaphore: sem_open failed");
     return ;
-  }
-  if (debug) {
-    int sval;
-    if (sem_getvalue(this->semaphore, &sval)) sval = -1;
-    printf("[posix-semaphore] Opened semaphore (initial value %u, current value %d)\n", value, sval);
   }
   this->locked = false;
   this->closed = false;
@@ -39,6 +27,7 @@ Semaphore::~Semaphore()
 
 void Semaphore::Init(v8::Local<v8::Object> exports)
 {
+  v8::Local<v8::Context> context = exports->CreationContext();
   Nan::HandleScope scope;
 
   // Prepare constructor template
@@ -51,18 +40,24 @@ void Semaphore::Init(v8::Local<v8::Object> exports)
   Nan::SetPrototypeMethod(tpl, "release", Release);
   Nan::SetPrototypeMethod(tpl, "close", Close);
 
-  constructor.Reset(tpl->GetFunction());
-  exports->Set(Nan::New("Semaphore").ToLocalChecked(), tpl->GetFunction());
+  constructor.Reset(tpl->GetFunction(context).ToLocalChecked());
+  exports->Set(
+    context,
+    Nan::New("Semaphore").ToLocalChecked(),
+    tpl->GetFunction(context).ToLocalChecked()
+  );
 }
 
 void Semaphore::New(const Nan::FunctionCallbackInfo<v8::Value>& info)
 {
-  bool    strict;
-  bool    debug;
-  bool    silent;
-  bool    retry_on_eintr;
-  char    *buf;
-  unsigned int     value;
+  v8::Isolate* isolate = info.GetIsolate();
+  v8::Local<v8::Context> context = isolate->GetCurrentContext();
+  bool strict;
+  bool debug;
+  bool silent;
+  bool retry_on_eintr;
+  char buf[SEMSIZE];
+  unsigned int value;
 
   if (!info.IsConstructCall())
     return Nan::ThrowError("Must call Semaphore() with new");
@@ -80,40 +75,34 @@ void Semaphore::New(const Nan::FunctionCallbackInfo<v8::Value>& info)
     return Nan::ThrowError("Semaphore() expects a boolean as fifth argument");
   if (!info[5]->IsUndefined() && !info[5]->IsUint32())
     return Nan::ThrowError("Semaphore() expects an integer as sixth argument");
-  strict = info[1]->BooleanValue();
-  debug = info[2]->BooleanValue();
-  silent = info[3]->BooleanValue();
-  retry_on_eintr = info[4]->BooleanValue();
-  value = !info[5]->IsUndefined()? info[5]->IntegerValue(): 1;
-  v8::String::Utf8Value str(info[0]->ToString());
-  if (str.length() >= 255 || str.length() <= 0)
+  strict = Nan::To<bool>(info[1]).FromJust();
+  debug = Nan::To<bool>(info[1]).FromJust();
+  silent = Nan::To<bool>(info[1]).FromJust();
+  retry_on_eintr = Nan::To<bool>(info[1]).FromJust();
+  value = !info[5]->IsUndefined()? info[5]->IntegerValue(context).FromJust(): 1;
+  v8::String::Utf8Value v8str(isolate, info[0]);
+  std::string str(*v8str);
+  strcpy(buf, str.c_str());
+
+  size_t str_len;
+  str_len = str.length();
+  if (str_len >= SEMSIZE - 1 || str_len <= 0)
     return Nan::ThrowError("Semaphore() : first argument's length must be < 255 && > 0");
-  buf = (char*)(*str);
-  Semaphore* obj = new Semaphore(buf, str.length(), strict, debug, silent, retry_on_eintr, value);
+
+  Semaphore* obj = new Semaphore(buf, str_len, strict, debug, silent, retry_on_eintr, value);
   obj->Wrap(info.This());
   info.GetReturnValue().Set(info.This());
 }
 
 void Semaphore::Acquire(const Nan::FunctionCallbackInfo<v8::Value>& info)
 {
-  int        r;
+  int r;
   Semaphore* obj = ObjectWrap::Unwrap<Semaphore>(info.Holder());
 
-  /*if (!obj->strict && (obj->closed || obj->locked))
-  {
-    if (obj->debug)
-      printf("[posix-semaphore] 'acquire' called when semaphore was already acquired or closed, but strict mode deactivated, so not failing\n");
-    return ;
-  }*/
   if (obj->strict && obj->closed)
     return Nan::ThrowError("trying to do operation over semaphore, but already closed");
   if (obj->strict && obj->locked)
     return Nan::ThrowError("trying to acquire semaphore, but already acquired");
-  if (obj->strict && obj->debug) {
-    int sval;
-    if (sem_getvalue(obj->semaphore, &sval)) sval = -1;
-    printf("[posix-semaphore] Before sem_wait, value = %d\n", sval);
-  }
   while ((r = sem_wait(obj->semaphore)) == -1 && errno == EINTR && obj->retry_on_eintr)
   {
     if (obj->debug)
@@ -139,59 +128,32 @@ void Semaphore::Acquire(const Nan::FunctionCallbackInfo<v8::Value>& info)
     return ;
   }
   obj->locked = true;
-  if (obj->debug) {
-    int sval;
-    if (sem_getvalue(obj->semaphore, &sval)) sval = -1;
-    printf("[posix-semaphore] After sem_wait, value = %d\n", sval);
-  }
 }
 
 void Semaphore::Release(const Nan::FunctionCallbackInfo<v8::Value>& info)
 {
   Semaphore* obj = ObjectWrap::Unwrap<Semaphore>(info.Holder());
-  
-  /*if (!obj->strict && (obj->closed || !obj->locked))
-  {
-    if (obj->debug)
-      printf("[posix-semaphore] 'release' called when semaphore was already released or closed, but strict mode deactivated, so not failing\n");
-    return ;
-  }*/
+
   if (obj->strict && obj->closed)
     return Nan::ThrowError("trying to do operation over semaphore, but already closed");
   if (obj->strict && !obj->locked)
     return Nan::ThrowError("trying to release semaphore, but already released");
-  if (obj->debug) {
-    int sval;
-    if (sem_getvalue(obj->semaphore, &sval)) sval = -1;
-    printf("[posix-semaphore] Before sem_post, value = %d\n", sval);
-  }
   if (sem_post(obj->semaphore) == -1)
   {
     if (obj->debug || !obj->silent)
     {
-      printf("[posix-semaphore] sem_post failed, printing errno message ('man sem_post' for more details on possible errors) : \n");
+      printf("[posix-semaphore] sem_post failed, printing errno message ('man sem_post' for more details on possible errors): \n");
       perror("[posix-semaphore] ");
     }
     return Nan::ThrowError("could not release semaphore, sem_post failed");
   }
   obj->locked = false;
-  if (obj->debug) {
-    int sval;
-    if (sem_getvalue(obj->semaphore, &sval)) sval = -1;
-    printf("[posix-semaphore] After sem_post, value = %d\n", sval);
-  }
 }
 
 void Semaphore::Close(const Nan::FunctionCallbackInfo<v8::Value>& info)
 {
   Semaphore* obj = ObjectWrap::Unwrap<Semaphore>(info.Holder());
 
-  /*if (!obj->strict && obj->closed)
-  {
-    if (obj->debug)
-      printf("[posix-semaphore] 'close' called when semaphore was already closed, but strict mode deactivated, so not failing\n");
-    return ;
-  }*/
   if (obj->strict && obj->closed)
     return Nan::ThrowError("trying to close semaphore, but already closed");
   if (obj->debug)

--- a/srcs/semaphore.cc
+++ b/srcs/semaphore.cc
@@ -76,9 +76,9 @@ void Semaphore::New(const Nan::FunctionCallbackInfo<v8::Value>& info)
   if (!info[5]->IsUndefined() && !info[5]->IsUint32())
     return Nan::ThrowError("Semaphore() expects an integer as sixth argument");
   strict = Nan::To<bool>(info[1]).FromJust();
-  debug = Nan::To<bool>(info[1]).FromJust();
-  silent = Nan::To<bool>(info[1]).FromJust();
-  retry_on_eintr = Nan::To<bool>(info[1]).FromJust();
+  debug = Nan::To<bool>(info[2]).FromJust();
+  silent = Nan::To<bool>(info[3]).FromJust();
+  retry_on_eintr = Nan::To<bool>(info[4]).FromJust();
   value = !info[5]->IsUndefined()? info[5]->IntegerValue(context).FromJust(): 1;
   v8::String::Utf8Value v8str(isolate, info[0]);
   std::string str(*v8str);

--- a/srcs/semaphore.cc
+++ b/srcs/semaphore.cc
@@ -82,10 +82,12 @@ void Semaphore::New(const Nan::FunctionCallbackInfo<v8::Value>& info)
   value = !info[5]->IsUndefined()? info[5]->IntegerValue(context).FromJust(): 1;
   v8::String::Utf8Value v8str(isolate, info[0]);
   std::string str(*v8str);
-  strcpy(buf, str.c_str());
 
   size_t str_len;
   str_len = str.length();
+  strncpy(buf, str.c_str(), str_len);
+  buf[str_len] = '\0';
+
   if (str_len >= SEMSIZE - 1 || str_len <= 0)
     return Nan::ThrowError("Semaphore() : first argument's length must be < 255 && > 0");
 

--- a/srcs/semaphore.h
+++ b/srcs/semaphore.h
@@ -1,5 +1,6 @@
 #ifndef SEMAPHORE_H
 #define SEMAPHORE_H
+#define SEMSIZE 256
 
 #include <nan.h>
 #include <fcntl.h>
@@ -19,9 +20,9 @@ class Semaphore : public Nan::ObjectWrap {
   static void Release(const Nan::FunctionCallbackInfo<v8::Value>& info);
   static void Close(const Nan::FunctionCallbackInfo<v8::Value>& info);
   static Nan::Persistent<v8::Function> constructor;
-  
+
   sem_t	*semaphore;
-  char	sem_name[256];
+  char	sem_name[SEMSIZE];
   bool	locked;
   bool	closed;
   bool  strict;

